Sources/Commands/SwiftPackageTool.swift

```diff
@@ -198,23 +214,165 @@ extension SwiftPackageTool {
+        var packageName: String
+
+        func run(_ swiftTool: SwiftTool) throws {
+            guard let cwd = localFileSystem.currentWorkingDirectory else {
+                throw InternalError("Could not find the current working directory.")
+            }
+            
+            guard let configPath = try swiftTool.getConfigPath() else {
+                throw InternalError("Could not find config path")
+            }
+            
+            guard !(packageType != nil && template != nil) else {
+                throw InternalError("Can't use --type in conjunction with --template")
+            }
+          
+            let initPackage = try InitPackage(fileSystem: localFileSystem,
+                                              configPath: configPath,
+                                              destinationPath: cwd.appending(component: packageName),
+                                              mode: .create,
+                                              packageName: packageName,
+                                              packageType: packageType,
+                                              packageTemplateName: template)
+            
             initPackage.progressReporter = { message in
                 print(message)
             }
-            try initPackage.writePackageStructure()
+            try initPackage.makePackage()
+        }
+    }
+    
+    struct AddTemplate: SwiftCommand {
+        static let configuration = CommandConfiguration(
+            abstract: "Donwloads/moves a template into the default template directory")
+        
+        @OptionGroup(_hiddenFromHelp: true)
+        var swiftOptions: SwiftToolOptions
+        
+        @Argument(help: "URL to template (absolute path to template works too)")
+        var url: String
+        
+        @Option(help: "Custom template name")
+        var name: String?
+
+        func run(_ swiftTool: SwiftTool) throws {
+            guard let configPath = try swiftTool.getConfigPath() else {
+                throw InternalError("Could not find config path")
+            }
+
+            let templatePath: AbsolutePath
+            if let path = try? AbsolutePath(validating: url) {
+                guard localFileSystem.exists(path) else {
+                    throw StringError("Could not find template: \(path)")
+                }
+                
+                guard localFileSystem.isDirectory(path) else {
+                    throw StringError("\(path) is not a valid directory")
+                }
+                
+                templatePath = configPath.appending(components: "templates", "new-package", name ?? path.basename)
+                try localFileSystem.copy(from: path, to: templatePath)
+            } else {
+                let provider = GitRepositoryProvider()
+                
+                if let templateName = name {
+                    templatePath = configPath.appending(components: "templates", "new-package", templateName)
```

Comment:
This appending operation is common throughout this PR. You could create an extension method on AbsolutePath and use that instead, e.g. 
```suggestion
                    templatePath = configPath.withTemplate(templateName)                    
```

E.g. 
```
extension AbsolutePath {
  func withTemplate(_ arg: String) {
      return appending(components: "templates", "new-package", arg)
  }
}
```